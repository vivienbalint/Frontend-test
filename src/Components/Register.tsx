import React, { useState } from "react";
import { useHistory } from "react-router";
import AuthContext from "../auth/auth.context";
import api from "../api/api.user";
import { Form, Button, Alert } from "react-bootstrap";

const Register = () => {
  let history = useHistory();
  let regex = /\[|]|"/g;
  const { dispatch } = React.useContext(AuthContext);

  const [{ username, email, password }, setRegisterData] = useState({
    username: "",
    email: "",
    password: "",
  });
  const [error, setError] = useState("");
  const [isSuccessful, setIsSuccessful] = useState(false);

  const handleRegister = async (e: React.FormEvent) => {
    e.preventDefault();
    const { user, error } = await api.createUser({
      username,
      email,
      password,
    });
    if (error) {
      setError(error);
      setIsSuccessful(false);
    } else if (user) {
      const userDataResponse = {
        userData: {
          email: user.email,
          username: user.username,
        },
        token: user.token,
      };
      setIsSuccessful(true);
      history.push("/user-page");
      dispatch({ type: "LOGIN", payload: userDataResponse });
    }
  };

  return (
    <div>
      {error.length > 0 && (
        <Alert variant={"danger"}>
          <Alert.Heading>Error:</Alert.Heading>
          <p>{error.replace(regex, "")}</p>
        </Alert>
      )}
      {isSuccessful && (
        <Alert variant={"success"}>
          <Alert.Heading>Success:</Alert.Heading>
          <p>You successfully registered yourself!</p>
        </Alert>
      )}
      <div className="box">
        <Form onSubmit={handleRegister} className="m-auto form-container">
          <h2 className="form-title">Register</h2>
          <Form.Group className="mb-3">
            <Form.Label column="lg">Username</Form.Label>
            <Form.Control
              size="lg"
              type="text"
              placeholder="Enter username"
              value={username}
              onChange={(e) =>
                setRegisterData({
                  username: e.target.value,
                  email,
                  password,
                })
              }
            />
          </Form.Group>
          <Form.Group className="mb-3">
            <Form.Label column="lg">Email address</Form.Label>
            <Form.Control
              size="lg"
              type="email"
              placeholder="Enter email address"
              value={email}
              onChange={(e) =>
                setRegisterData({
                  username,
                  email: e.target.value,
                  password,
                })
              }
            />
          </Form.Group>
          <Form.Group className="mb-3">
            <Form.Label column="lg">Password</Form.Label>
            <Form.Control
              size="lg"
              type="password"
              placeholder="Password"
              autoComplete="true"
              value={password}
              onChange={(e) =>
                setRegisterData({
                  username,
                  email,
                  password: e.target.value,
                })
              }
            />
          </Form.Group>
          <Button variant="primary" type="submit" size="lg">
            Register
          </Button>
        </Form>
      </div>
    </div>
  );
};

export default Register;
