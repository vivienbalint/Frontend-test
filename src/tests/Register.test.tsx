import React from "react";
import { mount, shallow } from "enzyme";
import Register from "../Components/Register";
import "../setupTests";

const dispatch = jest.fn();
React.useContext = (() => dispatch) as <T>(context: React.Context<T>) => T;

describe("Register", () => {
  const wrapper = mount(<Register />);

  it("should run without crashing", () => {
    shallow(<Register />);
  });

  it("should match the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should have a text field", () => {
    expect(wrapper.find('input[type="text"]').length).toEqual(1);
  });

  it("should have proper props for text field", () => {
    expect(wrapper.find('input[type="text"]').props()).toEqual({
      className: "form-control form-control-lg",
      id: undefined,
      readOnly: undefined,
      size: undefined,
      type: "text",
      placeholder: "Enter username",
      value: expect.any(String),
      onChange: expect.any(Function),
    });
  });

  it("should have an email field", () => {
    expect(wrapper.find('input[type="email"]').length).toEqual(1);
  });

  it("should have proper props for email field", () => {
    expect(wrapper.find('input[type="email"]').props()).toEqual({
      className: "form-control form-control-lg",
      id: undefined,
      readOnly: undefined,
      size: undefined,
      type: "email",
      placeholder: "Enter email address",
      value: expect.any(String),
      onChange: expect.any(Function),
    });
  });

  it("should have a password field", () => {
    expect(wrapper.find('input[type="password"]').length).toEqual(1);
  });

  it("should have proper props for password field", () => {
    expect(wrapper.find('input[type="password"]').props()).toEqual({
      className: "form-control form-control-lg",
      id: undefined,
      readOnly: undefined,
      size: undefined,
      type: "password",
      placeholder: "Password",
      value: expect.any(String),
      onChange: expect.any(Function),
      autoComplete: "true",
    });
  });

  it("should have a submit button", () => {
    expect(wrapper.find('button[type="submit"]').length).toEqual(1);
  });

  it("should have proper props for submit button", () => {
    expect(wrapper.find('button[type="submit"]').props()).toEqual({
      className: "btn btn-primary btn-lg",
      disabled: false,
      type: "submit",
      children: "Register",
    });
  });
});
