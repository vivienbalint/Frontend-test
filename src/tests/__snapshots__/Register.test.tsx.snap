// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register should match the snapshot 1`] = `"<div><div class=\\"box\\"><form class=\\"m-auto form-container\\"><h2 class=\\"form-title\\">Register</h2><div class=\\"mb-3\\"><label class=\\"form-label col-form-label col-form-label-lg col\\">Username</label><input placeholder=\\"Enter username\\" type=\\"text\\" class=\\"form-control form-control-lg\\" value=\\"\\"></div><div class=\\"mb-3\\"><label class=\\"form-label col-form-label col-form-label-lg col\\">Email address</label><input placeholder=\\"Enter email address\\" type=\\"email\\" class=\\"form-control form-control-lg\\" value=\\"\\"></div><div class=\\"mb-3\\"><label class=\\"form-label col-form-label col-form-label-lg col\\">Password</label><input placeholder=\\"Password\\" autocomplete=\\"true\\" type=\\"password\\" class=\\"form-control form-control-lg\\" value=\\"\\"></div><button type=\\"submit\\" class=\\"btn btn-primary btn-lg\\">Register</button></form></div></div>"`;
